README.md
# clamour

<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![Travis build status](https://travis-ci.com/lazappi/clamour.svg?branch=master)](https://travis-ci.com/lazappi/clamour)
[![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/lazappi/clamour?branch=master&svg=true)](https://ci.appveyor.com/project/lazappi/clamour)
[![Codecov test coverage](https://codecov.io/gh/lazappi/clamour/branch/master/graph/badge.svg)](https://codecov.io/gh/lazappi/clamour?branch=master)
<!-- badges: end -->

_clamour_ provides a template for setting up a website to display the results of
analysis of Twitter hashtags, more details are available on the
[documentation website][clamour].

You can see an example of the website produced by _clamour_ [here][example].

## Installation

You can install the current version of clamour from GitHub with:

``` r
# install.packages("remotes")
remotes::install_github("lazappi/clamour")
```

## Acknowledgements

Much of the analysis in the _clamour_ templates is originally based on work by
@neilws (and various tutorials on the internet) with contributions from
@jherrero. The _clamour_ package would not be possible without the contributions
of many developers to a range of packages, particularly _[usethis][]_ and
_[rtweet][]_.

## Code of conduct

Please note that the 'clamour' project is released with a
[Contributor Code of Conduct](CODE_OF_CONDUCT.md).
By contributing to this project, you agree to abide by its terms.


[usethis]: https://usethis.r-lib.org/ "usethis"
[rtweet]: https://rtweet.info/ "rtweet"
[clamour]: https://lazappi.github.io/clamour "clamour example"
[example]: https://lazappi.github.io/clamour-example "clamour example"

NEWS.md
# clamour 0.1.0 (2020-09-01)

* First minor version

# clamour 0.0.0

## clamour 0.0.0.9012 (2020-09-01)

* Add libraries script to templates
* Add knitr options script to templates
* Add `clamour_load_tweets()` function
* Add buttons to hide parameters and session info
* Improve formatting of topics tweets tables

## clamour 0.0.0.9011 (2020-09-01)

* Adapt analysis to allow multiple query hashtags (Fixes #3)
* Add plot of query hashtags
* Update `clamour_new()` to handle multiple hashtags
* Update introduction in hashtag template
* Limit number shown in hashtags plot

## clamour 0.0.0.9010 (2020-08-17)

* Add count of tweets and retweets to links table
* Shorten link url to save space
* Add regex for GitLab

## clamour 0.0.0.9009 (2020-08-17)

* Add analysis of other hashtags used (Fixes #4)

## clamour 0.0.0.9008 (2020-08-17)

* Add guards to tweet screenshots in template to catch when tweets have been
  deleted
* Add tweets by source plot (Fixes #2)

## clamour 0.0.0.9007 (2020-08-17)

* Turn off table of contents on index and about page templates (Fixes #1)

## clamour 0.0.0.9006 (2019-07-08)

* Add links to example website
* Add README to template

## clamour 0.0.0.9005 (2019-07-07)

* Add tweet shots to template analysis
* Update README and vignette
* Change pkgdown theme to sandstone

## clamour 0.0.0.9004 (2019-07-07)

* Add `clamour_new()` function
* Add `clamour_table()` and `clamour_list()` functions

## clamour 0.0.0.9003 (2019-07-06)

* Add example analysis files
* Add prompts to install packages and authorise rtweet

## clamour 0.0.0.9002 (2019-07-05)

* Start vignette

## clamour 0.0.0.9001 (2019-07-05)

* Add `clamour_setup()` function

## clamour 0.0.0.9000 (2019-07-05)

* Set up package structure


tests/testthat.R
library(testthat)
library(clamour)

test_check("clamour")


tests/testthat/test-setup.R
test_that("setup works", {
  path <- fs::path(tempdir(), "test")
  clamour_setup(path, rstudio = TRUE, open = FALSE)
  expect_true(fs::dir_exists(path))
  expect_true(fs::dir_exists(fs::path(path, "analysis")))
  expect_true(fs::file_exists(fs::path(path, "analysis", "_site.yml")))
  expect_true(fs::file_exists(fs::path(path, "analysis", "index.Rmd")))
  expect_true(fs::file_exists(fs::path(path, "analysis", "about.Rmd")))
  expect_true(fs::dir_exists(fs::path(path, "docs")))
  expect_true(fs::dir_exists(fs::path(path, "data")))
  fs::dir_delete(path)
})


R/new.R
#' New clamour analysis
#'
#' Creates a new analysis file based on the clamour template.
#'
#' @param name Name of the analysis. This is used to name files so should
#'   usually be short and simple.
#' @param hashtag The hashtag to use for this analysis, starting with "#". Can
#'   also be a vector of hashtags.
#' @param description A more detailed description of the analysis, details of
#'   the event etc.
#' @param start_day First day to be included in the analysis.
#' @param end_day Last day to be included in the analysis.
#' @param timezone Timezone to be used for the analysis.
#' @param theme Name of a [ggplot2](https://ggplot2.tidyverse.org/) theme
#'   function to use for plots.
#' @param accent Accent colour to use for plots.
#' @param accent2 Secondary accent colour to use for plots. If `NULL` this will
#'   be created based on `accent`.
#' @param kcore k-core to use for filtering the mentions network.
#' @param topics_k Number of topics to use for topic modelling.
#' @param bigram_filter Number of co-occurrences for filtering the bigram graph.
#' @param fixed Whether to continue to search and download new tweets. Set to
#'   `TRUE` to fix the dataset used for analysis.
#' @param seed Random seed set at the start of analysis.
#' @param dir Directory to save the new analysis file.
#'
#' @return Path to the newly created file, invisibly
#' @export
clamour_new <- function(name, hashtag,
                        description = paste("An analysis of", hashtag),
                        start_day = Sys.Date(), end_day = Sys.Date() + 1,
                        timezone = Sys.timezone(), theme = "theme_light",
                        accent = "dodgerblue", accent2 = NULL, kcore = 2,
                        topics_k = 6, bigram_filter = 3,
                        fixed = FALSE, seed = 1, dir = "analysis") {

    if (!all(startsWith(hashtag, "#"))) {
        usethis::ui_stop(
            "hashtag should start with #: {usethis::ui_value(hashtag)}"
        )
    }

    if (is.null(accent2)) {
        accent2 <- grDevices::rgb(
            t((255 - (255 - grDevices::col2rgb(accent)[, 1]) / 2)),
            maxColorValue = 255
        )
    }

    if (!fs::is_dir(usethis::proj_path(dir))) {
        usethis::ui_stop("{usethis::ui_path(dir)} is not a directory.")
    }

    name_path <- gsub("[^[:alnum:] ]", "", name)
    name_path <- gsub(" ", "_", name_path)
    name_path <- paste0(name_path, ".Rmd")

    data_list = list(
        TITLE       = name,
        DESCRIPTION = description,
        HASHTAG     = hashtag,
        START_DAY   = start_day,
        END_DAY     = end_day,
        TIMEZONE    = timezone,
        THEME       = theme,
        ACCENT      = accent,
        ACCENT2     = accent2,
        KCORE       = kcore,
        TOPICS      = topics_k,
        BIGRAM      = bigram_filter,
        FIXED       = fixed,
        SEED        = seed
    )

    usethis::use_template("hashtag_template.Rmd", fs::path(dir, name_path),
                          package = "clamour", data = data_list)

    invisible(fs::path(dir, name_path))
}

R/list.R
#' clamour table
#'
#' Read the YAML front matter from a set of clamour analysis files and produce
#' a tidy data.frame.
#'
#' @param path Path to directory to search for `.Rmd` files
#' @param ignore Vector or files to ignore
#'
#' @return data.frame containing information on analysis files
#'
#' @export
clamour_table <- function(path = ".",
                          ignore = c("index.Rmd", "about.Rmd",
                                     "_analysis.Rmd")) {

    rmds <- fs::dir_ls(path, glob = "*.Rmd")
    rmds <- rmds[!(rmds %in% ignore)]
    rmd_paths <- fs::path(path, rmds)

    yamls <- lapply(rmd_paths, rmarkdown::yaml_front_matter)

    rows <- lapply(seq_len(length(yamls)), function(idx) {
        yaml <- yamls[[idx]]
        rmd <- rmds[idx]
        data.frame(
            File        = rmd,
            Name        = yaml$title,
            Description = yaml$subtitle,
            Hashtag     = yaml$params$hashtag,
            StartDay    = as.Date(yaml$params$start_day),
            EndDay      = as.Date(yaml$params$end_day),
            Timezone    = yaml$params$timezone,
            Theme       = yaml$params$theme,
            Accent      = yaml$params$accent,
            Accent2     = yaml$params$accent2,
            kCore       = yaml$params$kcore,
            Topics      = yaml$params$topics_k,
            Bigram      = yaml$params$bigram_filter,
            Fixed       = yaml$params$fixed,
            Seed        = yaml$params$seed,

            stringsAsFactors = FALSE
        )
    })

    df <- do.call("rbind", rows)
    rownames(df) <- NULL

    return(df)
}

#' clamour list
#'
#' Produce a Markdown string with details of all the clamour analysis files in a
#' directory. This function is primarily designed to be used in the index of a
#' website so the list is automatically updated.
#'
#' @param path Path to directory to search for `.Rmd` files
#' @param ignore Vector or files to ignore
#'
#' @return Markdown string describing analysis files
#'
#' @export
clamour_list <- function(path = ".",
                         ignore = c("index.Rmd", "about.Rmd",
                                    "_analysis.Rmd")) {

    analyses <- clamour_table(path = path, ignore = ignore)

    str <- ""

    for (idx in seq_len(nrow(analyses))) {
        analysis <- analyses[idx, ]

        html <- gsub(".Rmd", ".html", analysis$File)

        str <- paste0(
            str,
            "* [", analysis$Name, "](", html, ")",
            " - ", analysis$Description,
            " (", analysis$StartDay, " - ", analysis$EndDay, ")",
            "\n"
        )
    }

    return(str)
}

R/load-tweets.R
#' Load tweets
#'
#' Load tweets for a clamour analysis
#'
#' @param query Query string to use for collecting new tweets.
#' @param cache_file Path to an `.Rds` file to store cached tweets.
#' @param timezone Time zone associated with tweets.
#' @param fixed If `TRUE` only read tweets from cache and do not download new
#'   tweets.
#'
#' @details
#' If a cache file exists tweets will first be read from here. New tweets will
#' also be downloaded if `fixed = FALSE`. Cached and new tweets are combined and
#' only the most recent version of each tweet is kept. Additional `datetime`,
#' `date` and `hour` fields are created using the supplied `timezone`.
#'
#' @return tibble containing tweets
#' @export
clamour_load_tweets <- function(query, cache_file = NULL,
                                timezone = Sys.timezone(),
                                fixed = FALSE) {

    cache_exists <- !is.null(cache_file) & fs::file_exists(cache_file)

    if (fixed & !cache_exists) {
        stop_msg <- paste(
            usethis::ui_code("fixed = TRUE"),
            "but", usethis::ui_code("cache_file"), "does not exist:",
            usethis::ui_path(as.character(cache_file)) # Convert NULL to char
        )
        usethis::ui_stop(stop_msg)
    }

    if (cache_exists) {
        cached_tweets <- readRDS(cache_file)
        usethis::ui_done(paste(
            "Loaded tweets from", usethis::ui_path(cache_file)
        ))
        usethis::ui_info(paste(
            usethis::ui_value(nrow(cached_tweets)),
            "tweets loaded from cache")
        )
    }

    if (fixed) {
        tweets <- cached_tweets
    } else {
        usethis::ui_todo("Downloading new tweets...")
        new_tweets <- rtweet::search_tweets(query, 10000) %>%
            dplyr::mutate(collected_at = Sys.time())
        usethis::ui_info(paste(
            "Found", usethis::ui_value(nrow(new_tweets)), "tweets")
        )

        if (cache_exists) {
            tweets <- new_tweets %>%
                rbind(cached_tweets) %>%
                dplyr::group_by(.data$status_id) %>%
                dplyr::top_n(1, .data$collected_at) %>%
                dplyr::ungroup()
        } else {
            tweets <- new_tweets
        }
    }

    if (!is.null(cache_file)) {
        usethis::ui_todo("Saving tweets to cache...")
        saveRDS(tweets, cache_file)
        usethis::ui_done(paste("Cached saved to", usethis::ui_path(cache_file)))
    }

    tweets <- tweets %>%
        dplyr::mutate(
            date     = lubridate::as_date(.data$created_at, tz = timezone),
            datetime = lubridate::as_datetime(.data$created_at, tz = timezone),
            hour     = lubridate::hour(.data$datetime)
        )
    usethis::ui_info(paste(
        "Full dataset contains", usethis::ui_value(nrow(tweets)), "tweets"
    ))

    return(tweets)
}


R/clamour-setup.R
#' Setup a clamour project
#'
#' This function creates the directory structure necessary for analysing Twitter
#' hashtag activity and displaying the results on a website. It will also prompt
#' you to install packages required by the default analysis and to authorise
#' [rtweet](https://rtweet.info/).
#'
#' @param path A path for the project. If it exists it will be used. If it does
#'   exist it will be created, provided the parent path exists.
#' @param rstudio If `TRUE`, calls [usethis::use_rstudio()] to make the new
#'   project into an [RStudio
#'   Project](https://support.rstudio.com/hc/en-us/articles/200526207-Using-Projects).
#'   If `FALSE` a sentinel `.here` file is placed so that the directory can be
#'   recognized as a project by the [here](https://here.r-lib.org) or
#'   [rprojroot](https://rprojroot.r-lib.org) packages.
#' @param open If `TRUE`, [activates][usethis::proj_activate()] the new project:
#'
#'   * If using RStudio desktop, the package is opened in a new session.
#'   * If on RStudio server, the current RStudio project is activated.
#'   * Otherwise, the working directory and active project is changed.
#'
#' @references
#' Based on [usethis::create_package()](https://github.com/r-lib/usethis/blob/master/R/create.R)
#'
#' @return Path to the newly created project, invisibly.
#' @export
clamour_setup <- function(path, rstudio = rstudioapi::isAvailable(),
                          open = interactive()) {

    path <- fs::path_expand(path)
    dir <- fs::path_dir(path)

    if (!fs::file_exists(dir)) {
        usethis::ui_stop("Directory {usethis::ui_path(dir)} does not exist.")
    }
    if (!fs::is_dir(dir)) {
        usethis::ui_stop("{usethis::ui_path(dir)} is not a directory.")
    }

    if (fs::file_exists(path) && !fs::dir_exists(path)) {
        usethis::ui_stop(
            "{usethis::ui_path(path)} exists but is not a directory"
        )
    }

    if (fs::dir_exists(path)) {
        usethis::ui_done("Using existing directory {usethis::ui_path(path)}")
    } else {
        fs::dir_create(path, recurse = TRUE)
        usethis::ui_done("Creating {usethis::ui_path(path)}")
    }

    old_project <- usethis::proj_set(path, force = TRUE)
    on.exit(usethis::proj_set(old_project), add = TRUE)

    usethis::use_directory("analysis")
    usethis::use_directory("analysis/figures", ignore = TRUE)
    usethis::use_template("_site.yml", "analysis/_site.yml",
                          package = "clamour")
    usethis::use_template("index.Rmd", "analysis/index.Rmd",
                          package = "clamour")
    usethis::use_template("about.Rmd", "analysis/about.Rmd",
                          package = "clamour")
    usethis::use_template("_analysis.Rmd", "analysis/_analysis.Rmd",
                          package = "clamour")
    clamour_new(
        name          = "EXAMPLE",
        hashtag       = "#rstats",
        description   = "An example analysis",
        start_day     = "2019-07-04",
        end_day       = "2019-07-06",
        timezone      = "Australia/Melbourne",
        theme         = "theme_light",
        accent        = "dodgerblue",
        accent2       = NULL,
        kcore         = 2,
        topics_k      = 6,
        bigram_filter = 3,
        fixed         = TRUE,
        seed          = 1
    )

    usethis::use_directory("R")
    usethis::use_template("libraries.R", "R/libraries.R", package = "clamour")
    usethis::use_template("knitr-options.R", "R/knitr-options.R",
                          package = "clamour")

    usethis::use_directory("docs")
    usethis::use_directory("data")
    if (!fs::file_exists(usethis::proj_path("data/EXAMPLE.Rds"))) {
        fs::file_copy(fs::path_package("clamour", "templates/example_data.Rds"),
                      usethis::proj_path("data/EXAMPLE.Rds"))
        usethis::ui_done("Copying {usethis::ui_path('data/EXAMPLE.Rds')}")
    }

    if (rstudio) {
        usethis::use_rstudio()
    }

    usethis::use_template("README.md", "README.md", package = "clamour")

    libs <- readLines(fs::path_package("clamour",
                                       "templates/libraries.R"))
    libs <- libs[grep("library", libs)]
    libs <- gsub('library\\(\\"', "", libs)
    libs <- gsub('\\"\\)', "", libs)
    libs <- libs[!(libs %in% utils::installed.packages())]
    n_libs <- length(libs)

    if (n_libs >= 1 && interactive()) {
        usethis::ui_info(
            "{n_libs} packages used in the default analysis are not installed:"
        )
        usethis::ui_line(usethis::ui_value(libs))
        install <- usethis::ui_yeah("Do you want to install them now?")
        if (install) {
            utils::install.packages(libs)
        }
    }

    usethis::ui_todo(
        "Run the following code to make sure rtweet is authorised:"
    )
    rt_test <- "rt <- rtweet::search_tweets('#rstats', n = 1000)"
    usethis::ui_line("{usethis::ui_code(rt_test)}")

    if (open) {
        if (usethis::proj_activate(path)) {
            # Working directory/active project changed; so don't undo on exit
            on.exit()
        }
    }

    invisible(usethis::proj_get())
}